src/index.ts

```diff
@@ -1,163 +1,171 @@
-        }
-
-        previousPoint = point;
-        segmentStartSample = segmentEndSample;
-    }
-}
-
-function calculateSortedComponentsFromOutput() {
-    components.splice(0, components.length);
-
-    for (let i = 0; i < fftSize; i++) {
-        const x = output[2 * i], y = output[2 * i + 1];
-        components.push({
-            frequency: i < fftSize / 2 ? i : i - fftSize,
-            magnitude: magnitude(x, y) / fftSize,
-            phase: Math.atan2(y, x) });
-    }
-
-    components.sort((a, b) => b.magnitude - a.magnitude);
-}
-
-function redraw() {
-    context.setTransform(window.devicePixelRatio, 0, 0, window.devicePixelRatio, 0, 0);
-    context.clearRect(0, 0, canvas.clientWidth, canvas.clientHeight);
-
-    const closedPath = new Path2D(unclosedPath);
-    closedPath.closePath();
-    context.strokeStyle = 'black';
-    context.stroke(closedPath);
-
-    if (components.length > 0) {
-        context.beginPath();
-
-        let x = 0, y = 0;
-
-        for (let i = 0; i < components.length; i++) {
-            const component = components[i];
-            const angle = parameter * component.frequency + component.phase;
-            x += component.magnitude * Math.cos(angle);
-            y += component.magnitude * Math.sin(angle);
-            context.lineTo(x, y);
-        }
-
-        context.strokeStyle = 'red';
-        context.stroke();
-    }
-}
+import FFT from 'fft.js'
+
+const canvas = document.getElementById('canvas') as HTMLCanvasElement;
+const context = canvas.getContext('2d')!;
+
+const points = new Array<{ x: number, y: number, segmentLength: number }>();
+let unclosedLength = 0;
+let unclosedPath = new Path2D();
+
+const fftSize = 4096;
+const fft = new FFT(fftSize);
+const input = fft.createComplexArray() as number[];
+const output = fft.createComplexArray() as number[];
+const components = new Array<{ frequency: number, magnitude: number, phase: number }>();
+let parameter = 0;
+let complexity = 0;
+let hasCapture = false;
+
+function updateCanvasSize() {
+    canvas.width = window.devicePixelRatio * canvas.clientWidth;
+    canvas.height = window.devicePixelRatio * canvas.clientHeight;
+}
+
+window.addEventListener('resize', function () { updateCanvasSize(); redraw(); });
+updateCanvasSize();
+
+canvas.onmousedown = function (e) {
+    if (e.button === 0) {
+        hasCapture = true;
+        //canvas.setPointerCapture((e as PointerEvent).pointerId);
+        addPoint(e.offsetX, e.offsetY);
+    }
+};
+
+canvas.ontouchstart = canvas.ontouchmove = function (e) {
+    if (e.touches.length === 1) {
+        addPoint(e.changedTouches[0].clientX - canvas.offsetLeft, e.changedTouches[0].clientY - canvas.offsetTop);
+        e.preventDefault();
+    }
+};
+
+canvas.onmousemove = function (e) {
+    if (hasCapture) addPoint(e.offsetX, e.offsetY);
+};
+
+canvas.onmouseup = function (e) {
+    if (hasCapture) {
+        hasCapture = false;
+        //canvas.releasePointerCapture((e as PointerEvent).pointerId);

```

Comment:
I should make this work across browsers.

Reply:
Which browser does this work for?
https://developer.mozilla.org/en-US/docs/Web/API/PointerEvent - PointerEvent inherits from MouseEvent, thus casting (MouseEvent)e to a PointerEvent is likely to fail on any browser.

Reply:
Firefox is the only one I tested one, but it's WIP and intended to work on all browsers.